### ISSUE TITLE
WebRequest Task stays alive and holds Connection after TimeoutException thrown

### ISSUE THREAD

**Issue by joelmm** [FirelyTeam/fhir-net-api#659]
`Hl7.Fhir.Rest.WebRequestExtensions.GetResponseAsync` returns a task, which creates another internal task to make the webcall (line 100). If the given timeout TimeSpan passes, the outer task throws `TimeoutException`, causing that Task to complete (line 122). However, the inner Task making the actual webcall does not get cancelled at that time, and continues trying to make the call for an indefinite amount of time until the call completes or fails (e.g., because of a `SocketException`). 

During this indefinite amount of time, the Task continues to hold a connection open. This causes problems with exceeding the [connection limit](https://msdn.microsoft.com/en-us/library/system.net.servicepoint.connectionlimit(v=vs.110).aspx) and causing subsequent calls to fail. 

One way to fix this is to add `request.Abort()` just before the timeout throw, as follows: 

```
                    try
                    {
                        if (!t.Wait(timeout))
                        {
                            request.Abort();
                            throw new TimeoutException();
                        }
                    }
```

Unfortunately, this will block the thread until the request completes, meaning that timeout won't really work at all. Still, this is certainly preferable to leaving a rogue Task holding a connection open after control returns to the caller.

Note that this is possibly related to https://github.com/ewoutkramer/fhir-net-api/issues/480. HttpClient has built-in timeout (using cancellation tokens), so this should be automatically fixed by that feature.

**Comment by joelmm** [comment]
To make this simpler to reproduce, here is a test example that does it. This uses an unrouteable IP, triggering a timeout. It makes the call five times in a row, waiting until the Task completes each time. You can observe from the printouts that the number of connections in use grows with each call, as there is still an inner Task open using a connection after the outer Task completes.

```
    [Test]
    public void MModalFhirClient_HttpTimeoutNew()
    {
      // Change the maximum number of connections and the timeout to observe the effect
      ServicePointManager.DefaultConnectionLimit = 3;
      var seconds = 5;

      var url = new Uri("http://www.google.com:81/fhir");
      var sp = ServicePointManager.FindServicePoint(url);
      Console.WriteLine($"{sp.CurrentConnections} of {sp.ConnectionLimit} total connections are in use");

      // Using unrouteable ip address www.google.com:81 (see https://stackoverflow.com/questions/100841/artificially-create-a-connection-timeout-error).
      var fc = new FhirClient(url) { Timeout = seconds * 1000 };

      var stopwatch = new Stopwatch();
      stopwatch.Start();

      for (int i = 0; i < 5; i++)

      {
        try
        {
          fc.GetAsync(url).Wait();
        }
        catch (Exception e)
        {
          // Print time elapsed, number of connections in use, and inner exception type (outer exception is System.AggregateException.
          Console.WriteLine($"{stopwatch.Elapsed}: {sp.CurrentConnections} of {sp.ConnectionLimit} total connections are in use. Exception type: {e.InnerException?.GetType()}");
        }
      }

      stopwatch.Stop();
    }
```

Output:

```
0 of 3 total connections are in use
00:00:05.0574762: 1 of 3 total connections are in use. Exception type: System.TimeoutException
00:00:10.0601867: 2 of 3 total connections are in use. Exception type: System.TimeoutException
00:00:15.0624209: 3 of 3 total connections are in use. Exception type: System.TimeoutException
00:00:20.0636607: 3 of 3 total connections are in use. Exception type: System.TimeoutException
00:00:25.0649957: 3 of 3 total connections are in use. Exception type: System.TimeoutException
```

Also note that the following pops up in VS after the test appears to complete, indicating that there are still uncompleted tasks holding the process open:
![image](https://user-images.githubusercontent.com/6687114/43716568-92d3b518-9952-11e8-8a7b-65ae51a09f83.png)

**Comment by wmrutten** [comment]
Great find, thank you for the issue report!
The inner task is a wrapper for HttpWebRequest.BeginGetResponse, which does not honor specified request timeout. MSDN documents an example how to force such a request to abort on timeout, using a call to ThreadPool.RegisterWaitForSingleObject:
https://msdn.microsoft.com/en-us/library/system.net.httpwebrequest.begingetresponse.aspx
However I don't know if it's worth the trouble, considering we're planning to update from HttpRequest to HttpClient anyway. Maybe useful as a legacy fix?

**Comment by joelmm** [comment]
The library has been really useful, so glad to contribute in a small way!

I did see that solution you linked. Wondering what you think of my one-line fix suggestion above. According to my test results, it solves the problem of leaving a rogue task open. Would it be worth creating a pull request with that fix?

Also, agreed re: the fact that HttpClient will fix this. Do you have any idea when that feature is going to be released? The issue report doesn't seem to have any definitive ETA.

**Comment by ewoutkramer** [comment]
@joelmm - as it stands I am lacking the time to work on that issue, although I would love to include it.

**Comment by ewoutkramer** [comment]
I think this is now fixed by PR #897!
